### PR TITLE
[GSoC 2019] Adds integration with Protego robots.txt parser.

### PR DESCRIPTION
**Pull request by anubhavp28**
[For Google Summer of Code 2019] Make Scrapy support [Protego](https://github.com/anubhavp28/protego) robots.txt parser.

### CODE REVIEW

File: scrapy/robotstxt.py [hunk 1]
```diff
@@ -110,3 +110,32 @@ def allowed(self, url, user_agent):
...
+        try:
+            robotstxt_body = robotstxt_body.decode('utf-8')
+        except UnicodeDecodeError:
+            # If we found garbage or robots.txt in an encoding other than UTF-8, disregard it.
+            # Switch to 'allow all' state.
+            logger.warning("Failure while parsing robots.txt using %(parser)s."
+                           " File either contains garbage or is in an encoding other than UTF-8, treating it as an empty file.",
+                           {'parser': "ProtegoRobotParser"},
+                           exc_info=sys.exc_info(),
+                           extra={'spider': self.spider})
```
Comment (kmike) [inline scrapy/robotstxt.py]: given that this code is repeated several times, I think it makes sense to create a function for it

**Comment by Gallaecio** [comment]
Github references can now point to https://github.com/anubhavp28/protego

Note: I’m delaying a proper review until https://github.com/scrapy/protego/pull/1 is merged and a release with it published in PyPI.

File: scrapy/robotstxt.py [hunk 2]
```diff
@@ -40,17 +55,7 @@ class PythonRobotParser(RobotParser):
...
-        except UnicodeDecodeError:
-            # If we found garbage or robots.txt in an encoding other than UTF-8, disregard it.
-            # Switch to 'allow all' state.
-            logger.warning("Failure while parsing robots.txt using %(parser)s."
-                           " File either contains garbage or is in an encoding other than UTF-8, treating it as an empty file.",
-                           {'parser': "RobotFileParser"},
-                           exc_info=sys.exc_info(),
-                           extra={'spider': self.spider})
-            robotstxt_body = ''
+        robotstxt_body = decode_robotstxt(robotstxt_body, spider, True)
```
Comment (kmike) [inline scrapy/robotstxt.py]: could you please use keyword argument to pass boolean value (True), to make code more readable - here and in other places?

File: tests/test_robotstxt_interface.py [hunk 3]
```diff
@@ -140,3 +140,15 @@ def setUp(self):
 
     def test_length_based_precedence(self):
         raise unittest.SkipTest("Rerp does not support length based directives precedence.")
+
+
+class ProtegoRobotParserTest(BaseRobotParserTest, unittest.TestCase):
+    if not rerp_available():
```
Comment (kmike) [inline tests/test_robotstxt_interface.py]: I think it should check for protego, not for rerp

**Pull request merged**